core/network/impl/router_libp2p.cpp

```diff
@@ -111,24 +111,36 @@ namespace kagome::network {
 
   void RouterLibp2p::readGossipMessage(std::shared_ptr<Stream> stream) const {
     auto read_writer = std::make_shared<ScaleMessageReadWriter>(stream);
-    read_writer->read<GossipMessage>(
-        [self{shared_from_this()},
-         stream = std::move(stream)](auto &&msg_res) mutable {
-          if (!msg_res) {
-            self->log_->error("error while reading gossip message: {}",
-                              msg_res.error().message());
-            return stream->reset();
-          }
+    read_writer->read<GossipMessage>([wp = weak_from_this(),
+                                      stream = std::move(stream)](
+                                         auto &&msg_res) mutable {
+      auto self = wp.lock();
+      if (not self) return;
+
+      if (not msg_res) {
+        self->log_->error("error while reading gossip message: {}",
+                          msg_res.error().message());
+        return stream->reset();
+      }
 
-          if (!self->processGossipMessage(msg_res.value())) {
-            stream->reset();
-            return;
-          }
-          self->readGossipMessage(stream);
-        });
+      auto peer_id_res = stream->remotePeerId();
+      if (not peer_id_res.has_value()) {
+        self->log_->error("can't get peer_id for gossip message: {}",
+                          msg_res.error().message());
```

Comment:
Is it possible?

Reply:
It depends on implementation of stream. It's a "black box" for that code.